opslevel/provider.go

```diff
@@ -1,137 +1,278 @@
-			"opslevel_check_repository_grep":       resourceCheckRepositoryGrep(),
-			"opslevel_check_repository_integrated": resourceCheckRepositoryIntegrated(),
-			"opslevel_check_repository_search":     resourceCheckRepositorySearch(),
-			"opslevel_check_service_dependency":    resourceCheckServiceDependency(),
-			"opslevel_check_service_configuration": resourceCheckServiceConfiguration(),
-			"opslevel_check_service_ownership":     resourceCheckServiceOwnership(),
-			"opslevel_check_service_property":      resourceCheckServiceProperty(),
-			"opslevel_check_tag_defined":           resourceCheckTagDefined(),
-			"opslevel_check_tool_usage":            resourceCheckToolUsage(),
-			"opslevel_domain":                      resourceDomain(),
-			"opslevel_filter":                      resourceFilter(),
-			"opslevel_infrastructure":              resourceInfrastructure(),
-			"opslevel_integration_aws":             resourceIntegrationAWS(),
-			"opslevel_property_assignment":         resourcePropertyAssignment(),
-			"opslevel_property_definition":         resourcePropertyDefinition(),
-			"opslevel_repository":                  resourceRepository(),
-			"opslevel_rubric_level":                resourceRubricLevel(),
-			"opslevel_rubric_category":             resourceRubricCategory(),
-			"opslevel_scorecard":                   resourceScorecard(),
-			"opslevel_secret":                      resourceSecret(),
-			"opslevel_service":                     resourceService(),
-			"opslevel_service_dependency":          resourceServiceDependency(),
-			"opslevel_service_repository":          resourceServiceRepository(),
-			"opslevel_service_tag":                 resourceServiceTag(),
-			"opslevel_service_tool":                resourceServiceTool(),
-			"opslevel_system":                      resourceSystem(),
-			"opslevel_tag":                         resourceTag(),
-			"opslevel_team_contact":                resourceTeamContact(),
-			"opslevel_team_tag":                    resourceTeamTag(),
-			"opslevel_team":                        resourceTeam(),
-			"opslevel_trigger_definition":          resourceTriggerDefinition(),
-			"opslevel_user":                        resourceUser(),
-			"opslevel_webhook_action":              resourceWebhookAction(),
-		},
+func (p *OpslevelProvider) Configure(ctx context.Context, req provider.ConfigureRequest, resp *provider.ConfigureResponse) {
+	var data OpslevelProviderModel
+	defaultApiUrl := "https://api.opslevel.com/"
+	defaultApiTimeout := 10
+	tflog.Info(ctx, "Initializing opslevel client")
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
-		ConfigureFunc: func(d *schema.ResourceData) (interface{}, error) {
-			url := d.Get("api_url").(string)
-			token := d.Get("api_token").(string)
-			timeout := d.Get("api_timeout").(int)
-			if timeout <= 0 {
-				timeout = 10
+	// Configuration values are now available.
```

Comment:
the following is a REALLY large function - is it possible to componentize this into smaller functions similar to the "functional options" concept in go.  Could even make it more generic so if the function returns a golang error you just feed that into `resp.Diagnostics.AddError(err.Message())`

Reply:
Of course. I just had not yet prioritized this refactor work